### PR TITLE
AEA-1325 - fix/dhtpeer message ordering

### PR DESCRIPTION
**Pull request by siddharthantier**
## Fixes

Fixing communications between peer and client in an ordered/linear manner

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 1]
```diff
@@ -787,6 +792,7 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
 				}
 			}
 		}()
+		wg.Wait()
```
Comment (lrahmani) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:795]: this is functionally equivalent to not creating the goroutine, i.e., reduces the degree of parallelism to 1

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go [hunk 2]
```diff
@@ -758,6 +758,67 @@ func TestRoutingDelegateClientToDHTPeerIndirect(t *testing.T) {
 	expectEnvelope(t, client.Rx)
 }
 
+// TestMessageOrderingWithDelegateClient
+func TestMessageOrderingWithDelegateClient(t *testing.T) {
```
Comment (lrahmani) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:762]: This doesn't test the message ordering at reception

**Review by lrahmani — CHANGES_REQUESTED** [review]
LGTM with caveats:

- the solution will work but it completely avoids the problem, which is fine
- the test at it's current form is not useful

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 3]
```diff
@@ -769,24 +776,47 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
...
-			} else {
-				err := dhtPeer.RouteEnvelope(envel)
-				if err != nil {
-					lerror(err).Str("addr", addr).
-						Msg("while routing delegate client envelope")
-					// TODO() send error back
+		// initializing pairwise channel queues and one go routine per pair
+		pair := envel.To + envel.Sender
+		if _, ok := dhtPeer.syncMessages[pair]; !ok {
+			dhtPeer.syncMessages[pair] = make(chan *aea.Envelope, 100)
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:789]: The buffer value is somewhat arbitrary. It needs to be sufficiently high to allow for some buildup.

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go (hunk 2 above)
Comment (DavidMinarsch) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:764]: It does now, have a look

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 4]
```diff
@@ -769,24 +776,47 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
...
+						lerror(err).Str("addr", addr).
+							Msg("while routing delegate client envelope")
+						// TODO() send error back
+					}
 				}
-			}
-		}()
+			}()
+		}
+		// add to queue (nonblocking - buffered queue)
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:808]: General concept: one producer, multiple consumers (go-routines), one for each pair. This way we ensure we have linerarity for each pair and concurrency across pairs.

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go (hunk 3 above)
Comment (lrahmani) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:792]: 100 is small, bump it to atleast a 1000

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 5]
```diff
@@ -769,24 +779,44 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
...
-			} else {
-				err := dhtPeer.RouteEnvelope(envel)
-				if err != nil {
-					lerror(err).Str("addr", addr).
-						Msg("while routing delegate client envelope")
-					// TODO() send error back
+		// initializing pairwise channel queues and one go routine per pair
+		pair := envel.To + envel.Sender
+		if _, ok := dhtPeer.syncMessages[pair]; !ok {
+			dhtPeer.syncMessages[pair] = make(chan *aea.Envelope, 100)
```
Comment (lrahmani) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:792]:
```suggestion
			dhtPeer.syncMessages[pair] = make(chan *aea.Envelope, 1000)
```

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go (hunk 4 above)
Comment (lrahmani) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:811]: nice!

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go [hunk 6]
```diff
@@ -1723,6 +1821,20 @@ func expectEnvelope(t *testing.T, rx chan *aea.Envelope) {
 	}
 }
 
+func expectEnvelopeOrdered(t *testing.T, rx chan *aea.Envelope, counter int) {
```
Comment (lrahmani) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:1824]: you can reuse `expectEnvelope` by making it return the envelope

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 7]
```diff
@@ -143,6 +143,9 @@ type DHTPeer struct {
 	closing    chan struct{}
```
Comment (lrahmani) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:143]: @DavidMinarsch are you planning to keep two copies of the `libp2p_node` project?

File: packages/fetchai/connections/p2p_libp2p/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 8]
```diff
@@ -769,24 +779,44 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
...
-			} else {
-				err := dhtPeer.RouteEnvelope(envel)
-				if err != nil {
-					lerror(err).Str("addr", addr).
-						Msg("while routing delegate client envelope")
-					// TODO() send error back
+		// initializing pairwise channel queues and one go routine per pair
+		pair := envel.To + envel.Sender
+		if _, ok := dhtPeer.syncMessages[pair]; !ok {
+			dhtPeer.syncMessages[pair] = make(chan *aea.Envelope, 100)
```
Comment (lrahmani) [inline packages/fetchai/connections/p2p_libp2p/libp2p_node/dht/dhtpeer/dhtpeer.go:792]:
```suggestion
			dhtPeer.syncMessages[pair] = make(chan *aea.Envelope, 1000)
```

**Review by lrahmani — APPROVED** [review]
Looks Neat!

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go (hunk 7 above)
Comment (DavidMinarsch) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:143]: Not permanently. But for now this is easier from a package management pov

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**